File: CombinationSum.java
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;


public class CombinationSum {
    public static List<List<Integer>> combinationSum(int[] candidates, int target) {
        List<List<Integer>> result = new ArrayList<List<Integer>>();
        List<Integer> comb = new ArrayList<Integer>();
         int[] temp = Arrays.copyOf(candidates, candidates.length);
         Arrays.sort(temp);
        combinationSum(temp, target, 0, result, comb);
        return result;
    }
    private static void combinationSum(int[] candidates, int target, int start, List<List<Integer>> result, List<Integer>comb) {
        if (target == 0) {
            if (comb.size() != 0) {
                List<Integer> temp = new ArrayList<Integer>(comb);
                result.add(temp);
            }
            return;
        }
        if (target < 0 || start == candidates.length)
            return;
        for (int i = start; i < candidates.length; i++) {
            comb.add(candidates[i]);
            combinationSum(candidates, target - candidates[i], i, result, comb);
            comb.remove(comb.size() - 1);
        }
    }
    public static void main(String[] args) {
        int[] nums = {8, 7, 4, 3};
        int n = 11;
        List<List<Integer>> result = combinationSum(nums, n);
        for (List<Integer> list: result) {
            for (Integer i: list) {
                System.out.print(" " + i);
            }
            System.out.println("");
        }
    }
}

File: py_solution/combination_sum.py
class Solution(object):
    def combinationSum(self, candidates, target):
        """
        :type candidates: List[int]
        :type target: int
        :rtype: List[List[int]]
        """
        res = []
        nums = []
        s = [i for i in candidates]
        s.sort()
        self._combination_sum(s, 0, target, nums, res)
        return res

    def _combination_sum(self, candidates, start, remain, nums, res):
        if remain < 0 or start == len(candidates):
            return
        if remain == 0:
            res.append([i for i in nums])
            return
        for i in range(start, len(candidates)):
            if candidates[i] > remain:
                break
            nums.append(candidates[i])
            self._combination_sum(candidates, i, remain - candidates[i], nums, res)
            nums.pop()


if __name__ == "__main__":
    solution = Solution()
    assert solution.combinationSum([2,3,6,7], 7) == [[2, 2, 3], [7]]
    assert solution.combinationSum([2,3,5], 8) == [[2,2,2,2], [2,3,3], [3,5]]
    assert solution.combinationSum([8,7,4,3], 11) == [[3,4,4],[3,8],[4,7]]